connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java

```diff
@@ -0,0 +1,155 @@
+                        Driver.class,
+                        loader
+                );
+                Iterator<Driver> driversIterator = loadedDrivers.iterator();
+                try {
+                    while (driversIterator.hasNext()) {
+                        Driver driver = driversIterator.next();
+                        log.debug(
+                                "Registered java.sql.Driver: {} to java.sql.DriverManager",
+                                driver
+                        );
+                    }
+                } catch (Throwable t) {
+                    log.debug(
+                            "Ignoring java.sql.Driver classes listed in resources but not"
+                                    + " present in class loader's classpath: ",
+                            t
+                    );
+                }
+                return null;
+            }
+        );
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    protected <T> PluginDesc<T> pluginDesc(Class<? extends T> plugin, String version, ClassLoader loader) {
+        return new PluginDesc(plugin, version, loader);
+    }
+
+    @SuppressWarnings("unchecked")
+    protected <T> SortedSet<PluginDesc<T>> getServiceLoaderPluginDesc(Class<T> klass, ClassLoader loader) {
+        SortedSet<PluginDesc<T>> result = new TreeSet<>();
+        ServiceLoader<T> serviceLoader = ServiceLoader.load(klass, loader);
+        for (Iterator<T> iterator = serviceLoader.iterator(); iterator.hasNext(); ) {
+            try (LoaderSwap loaderSwap = withClassLoader(loader)) {
+                T pluginImpl;
+                try {
+                    pluginImpl = iterator.next();
+                } catch (ServiceConfigurationError t) {
+                    log.error("Failed to discover {}{}", klass.getSimpleName(), reflectiveErrorDescription(t.getCause()), t);
+                    continue;
+                }
+                Class<? extends T> pluginKlass = (Class<? extends T>) pluginImpl.getClass();
+                if (pluginKlass.getClassLoader() != loader) {
+                    log.debug("{} from other classloader {} is visible from {}, excluding to prevent isolated loading",
+                            pluginKlass.getSimpleName(), pluginKlass.getClassLoader(), loader);
+                    continue;
+                }
+                result.add(pluginDesc(pluginKlass, versionFor(pluginImpl), loader));
+            }
+        }
+        return result;
+    }
+
+    protected static <T> String versionFor(T pluginImpl) {
+        try {
+            if (pluginImpl instanceof Versioned) {
+                return ((Versioned) pluginImpl).version();
+            }
+        } catch (Throwable t) {
+            log.error("Failed to get plugin version for " + pluginImpl.getClass(), t);
+        }
+        return DelegatingClassLoader.UNDEFINED_VERSION;
+    }
+
+    protected String reflectiveErrorDescription(Throwable t) {
```

Comment:
Nit: this was `static` before, which is a bit better since we never need to mock it and it demonstrates that the method doesn't rely on instance state.
```suggestion
    protected static String reflectiveErrorDescription(Throwable t) {
```